File: src/store.ts
/** In-memory task store with JSON persistence. */

import { existsSync, readFileSync, writeFileSync } from "node:fs";

export const VALID_STATES = ["todo", "doing", "done"] as const;

export type TaskState = (typeof VALID_STATES)[number];

export interface Task {
  id: number;
  title: string;
  state: TaskState;
}

/** Holds tasks and persists them as a JSON file. */
export class TaskStore {
  private tasks = new Map<number, Task>();
  private nextId = 1;
  private path?: string;

  constructor(path?: string) {
    this.path = path;
    if (this.path && existsSync(this.path)) {
      this.load();
    }
  }

  add(title: string): number {
    const id = this.nextId;
    this.tasks.set(id, { id, title, state: "todo" });
    this.nextId += 1;
    this.save();
    return id;
  }

  move(id: number, state: TaskState): void {
    if (!VALID_STATES.includes(state)) {
      throw new Error(`invalid state: ${state}`);
    }
    const task = this.tasks.get(id);
    if (!task) {
      throw new Error(`no such task: ${id}`);
    }
    task.state = state;
    this.save();
  }

  pending(): Task[] {
    return [...this.tasks.values()].filter((t) => t.state !== "done");
  }

  private load(): void {
    const data = JSON.parse(readFileSync(this.path!, "utf8"));
    this.tasks = new Map(data.tasks.map((t: Task) => [t.id, t]));
    this.nextId = data.nextId;
  }

  private save(): void {
    if (!this.path) return;
    const payload = { tasks: [...this.tasks.values()], nextId: this.nextId };
    writeFileSync(this.path, JSON.stringify(payload, null, 2));
  }
}

File: taskbox/store.py
"""In-memory task store with JSON persistence."""

import json
from pathlib import Path

VALID_STATES = ("todo", "doing", "done")


class TaskStore:
    """Holds tasks and persists them as a JSON file."""

    def __init__(self, path=None):
        self.path = Path(path) if path else None
        self.tasks = {}
        self.next_id = 1
        if self.path and self.path.exists():
            self._load()

    def add(self, title):
        task_id = self.next_id
        self.tasks[task_id] = {"id": task_id, "title": title, "state": "todo"}
        self.next_id += 1
        self._save()
        return task_id

    def move(self, task_id, state):
        if state not in VALID_STATES:
            raise ValueError(f"invalid state: {state}")
        if task_id not in self.tasks:
            raise KeyError(f"no such task: {task_id}")
        self.tasks[task_id]["state"] = state
        self._save()

    def pending(self):
        return [t for t in self.tasks.values() if t["state"] != "done"]

    def _load(self):
        data = json.loads(self.path.read_text())
        self.tasks = {t["id"]: t for t in data["tasks"]}
        self.next_id = data["next_id"]

    def _save(self):
        if not self.path:
            return
        payload = {"tasks": list(self.tasks.values()), "next_id": self.next_id}
        self.path.write_text(json.dumps(payload, indent=2))
